### PR TITLE
Add __traits isDynamicArray

### LINKED ISSUE #23784: Add is-expression or `__traits` for Dynamic Arrays
@WalterBright As discussed at DConf, we need to add an is-expression or a `__traits` for dynamic arrays.

### PR DESCRIPTION
**Pull request by nordlow**
Partially resolves https://github.com/dlang/dmd/issues/23784.

~~I can split the PR up into two if requested.~~

Changelog will be updated if this gets approved.

~~Looking into failing tests.~~

### CODE REVIEW

**Comment by ntrel** [comment]
`runnable/traits.d` TEST_OUTPUT needs updating.

**Comment by dkorpel** [comment]
What's the rationale for isArray and why is an Associative Array not one?

**Comment by dkorpel** [comment]
From the issue

> Please do not allow this trait to accept enums!

This PR's implementation accepts enums

**Comment by nordlow** [comment]
> What's the rationale for isArray and why is an Associative Array not one?

Good point. I agree that isArray should be true for AAs.

Update: I removed isArray from this PR.

**Comment by nordlow** [comment]
> From the issue
> 
> > Please do not allow this trait to accept enums!
> 
> This PR's implementation accepts enums

Shouldn't we mimic the behavior `__traits(isStaticArray, ...)` in this case? Despite it accepting `enum`s is a bad design.

And then later disallow all traits of this kind to not accept enums?

For the sake of consistency.

I'm ok with both alternatives.

**Comment by dkorpel** [comment]
> Good point. I agree that isArray should be true for AAs.

That's not what I meant, I don't see why it needs to exist in the first place. What's the real world use case?

**Comment by nordlow** [comment]
> > Good point. I agree that isArray should be true for AAs.
> 
> That's not what I meant, I don't see why it needs to exist in the first place. What's the real world use case?

I don't know.

**Comment by nordlow** [comment]
`compiler/test/run.d runnable/traits.d` now passes locally for me.

**Comment by nordlow** [comment]
> What's the rationale for isArray and why is an Associative Array not one?

Removed `isArray` changes from this PR.

File: changelog/dmd.traits.isDynamicArray.dd [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Add a new unary builtin trait `isDynamicArray` that evalutes to true
+iff its single type argument is a D (fat-pointer) array (slice) of any
+type.
```
Comment (thewilsonator) [inline changelog/dmd.traits.isDynamicArray.dd]: newline at end of file missing

Comment (nordlow) [reply changelog/dmd.traits.isDynamicArray.dd]: Done.

**Comment by Geod24** [comment]
Can't we do this with an `is` expression ?

**Comment by nordlow** [comment]
> Can't we do this with an `is` expression ?

Yes, as seen in

```d
template isDynamicArray(T)
{
    static if (is(T == U[], U))
        enum bool isDynamicArray = true;
    else static if (is(T U == enum))
        // BUG: isDynamicArray / isStaticArray considers enums
        // with appropriate base types as dynamic/static arrays
        // Retain old behaviour for now, see
        // https://github.com/dlang/phobos/pull/7574
        enum bool isDynamicArray = isDynamicArray!U;
    else
        enum bool isDynamicArray = false;
}
```

So this addition is for conformity with the existing builtin traits `isStaticArray` and `isAssociativeArray`.

The behavior of supporting enums for all the builtin traits `is...Array` might not be the preferred default. But we copy the behavior of the builtin `isStaticArray` for, again, conformity.

Note, that long-term I would like increase the convenience of builtin traits by somehow making them referreable to as identifiers and passable as alias parameters like how traits in `std.meta` and `std.traits` are currently used.

**Comment by nordlow** [comment]
DAutoTest failure seems like an unrelated timeout to me.

**Comment by kinke** [comment]
> object.Exception@../tools/changed.d(556): Changelog entries should consist of one title line, a blank separator line, and a description.

I guess it doesn't like that your changelog file starts with an empty line.

File: compiler/test/runnable/traits.d [hunk 2]
```diff
@@ -265,6 +271,36 @@ void test7()
     assert(__traits(isStaticArray, dchar) == false);
 }
 
+void test7d()
+{
+    assert(__traits(isDynamicArray) == false);
+    assert(__traits(isDynamicArray, S) == false);
+    assert(__traits(isDynamicArray, C) == false);
+    assert(__traits(isDynamicArray, E) == false);
+    assert(__traits(isDynamicArray, void*) == false);
```
Comment (kinke) [inline compiler/test/runnable/traits.d:284]: A missing interesting case is an enum with slice basetype (e.g., `enum E : string`) - where the implementation currently yields true AFAICT.

**Edit**: Ah, I've missed the discussion above. Anyway, that edge case definitely deserves a test case.

Comment (nordlow) [reply compiler/test/runnable/traits.d:284]: Test case added.

Comment (nordlow) [reply compiler/test/runnable/traits.d:284]: Shall I include a test case for

```d
enum ES2 : char[2] { a = "aa", b = "bb", c = "cc" }
assert(__traits(isStaticArray, ES2) == true);
```

?

**Review by ntrel — CHANGES_REQUESTED** [review]
This needs an update to `spec/traits.dd`.

**Comment by nordlow** [comment]
> This needs an update to `spec/traits.dd`.

Added.

**Review by ntrel — APPROVED** [review]

**Comment by nordlow** [comment]
Anything else? If not, ready for merge.

**Comment by dkorpel** [comment]
Perhaps a quick check from @LightBender and @WalterBright since their DConf conversation sparked the issue.
